[DDP] Mark a few variables as const in reducer

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#54764 [DDP] Mark a few variables as const in reducer**
* #54763 [c10d] s/torch::autograd::variable/at::Tensor/g

We mark a few vars as const in Reducer, also do this for replicas_ and
process_group_ as they should not be changed by Reducer during training.

Differential Revision: [D27357132](https://our.internmc.facebook.com/intern/diff/D27357132/)

rohan-varma committed:
[DDP] Mark a few variables as const in reducer

We mark a few vars as const in Reducer, also do this for replicas_ and
process_group_ as they should not be changed by Reducer during training. This
can help eliminate issues at compile time and prevent the developer from
accidently changing these variables.

Differential Revision: [D27357132](https://our.internmc.facebook.com/intern/diff/D27357132/)

[ghstack-poisoned]

wayi1 reviewed (approved):
Thanks for improving the code quality!

rohan-varma committed:
Update on "[DDP] Mark a few variables as const in reducer"


We mark a few vars as const in Reducer, also do this for replicas_ and
process_group_ as they should not be changed by Reducer during training.

Differential Revision: [D27357132](https://our.internmc.facebook.com/intern/diff/D27357132/)

[ghstack-poisoned]
